File: app/assets/javascripts/MainController.js
(function () {
    var app = angular.module('shocktrade');

    /**
     * Main Controller
     * @author lawrence.daniels@gmail.com
     */
    app.controller('MainController', ['$scope', '$http', '$interval', '$location', '$log', '$timeout', 'toaster', 'ContestService', 'Facebook', 'FavoriteSymbols', 'HeldSecurities', 'MarketStatus', 'MySession', 'ProfileService', 'RecentSymbols', 'SignUpDialog',
        function ($scope, $http, $interval, $location, $log, $timeout, toaster, ContestService, Facebook, FavoriteSymbols, HeldSecurities, MarketStatus, MySession, ProfileService, RecentSymbols, SignUpDialog) {
            // setup the loading mechanism
            $scope._loading = false;
            $scope.loading = false;
            $scope.admin = false;

            // mapping of online players
            var onlinePlayers = {};

            ///////////////////////////////////////////////////////////////////////////
            //          Public Functions
            ///////////////////////////////////////////////////////////////////////////

            $scope.mainInit = function(uuid) {
                console.log("Session UUID is " + uuid);
            };

            $scope.abs = function (value) {
                return !value ? value : ((value < 0) ? -value : value);
            };

            $scope.getDate = function(date) {
                return date && date.$date ? date.$date : date;
            };

            $scope.getHtmlQuote = function (q) {
                if(!q) return "";
                else {
                    var html = "<i class='" + $scope.getAssetIcon(q) + "'></i> " + q.symbol + ' - ' + q.name;
                    console.log("html = " + html);
                    return html;
                }
            };

            $scope.getAssetCode = function (q) {
                if (!q) return null;
                else {
                    switch (q.assetType) {
                        case 'Crypto-Currency':
                            return "&#xf15a;"; // fa-bitcoin
                        case 'Currency':
                            return "&#xf155;"; // fa-dollar
                        case 'ETF':
                            return "&#xf18d;"; // fa-stack-exchange
                        default:
                            return "&#xf0ac;"; // fa-globe
                    }
                }
            };

            $scope.getAssetIcon = function (q) {
                if (!q || !q.assetType) return "fa fa-globe st_blue";
                else {
                    switch (q.assetType) {
                        case 'Crypto-Currency':
                            return "fa fa-bitcoin st_blue";
                        case 'Currency':
                            return "fa fa-dollar st_blue";
                        case 'ETF':
                            return "fa fa-stack-exchange st_blue";
                        default:
                            return "fa fa-globe st_blue";
                    }
                }
            };

            /**
             * Returns the contacts matching the given search term
             */
            $scope.getRegisteredFriends = function () {
                return MySession.fbFriends;
            };

            $scope.getBarRanking = function () {
                if (MySession.contestIsEmpty() || !MySession.getUserName()) return null;
                else {
                    var rankings = ContestService.getPlayerRankings(MySession.getContest(), MySession.getUserName());
                    return rankings.player;
                }
            };

            $scope.changePlayTab = function (tabIndex) {
                var tab = $scope.playTabs[tabIndex];
                $log.info("Changing location to " + tab.url);
                $location.path(tab.url);
                return true;
            };

            $scope.getTabIndex = function () {
                return determineTableIndex();
            };

            function determineTableIndex() {
                var path = $location.path();
                if (path.indexOf("/connect") != -1) return 1;
                else if (path.indexOf("/discover") != -1) return 2;
                else if (path.indexOf("/explore") != -1) return 2;
                else if (path.indexOf("/dashboard") != -1) return 1;
                else if (path.indexOf("/news") != -1) return 2;
                else if (path.indexOf("/research") != -1) return 3;
                else if (path.indexOf("/search") != -1) return 0;
                else if (path.indexOf("/profile") != -1) return 4;
                else return 0;
            }

            $scope.isLoading = function () {
                return $scope._loading;
            };

            $scope.startLoading = function (timeout) {
                $scope._loading = true;
                /*
                 var _timeout = timeout || 4000;

                 // set loading timeout
                 var promise = $timeout(function() {
                 console.log("Disabling the loading animation due to time-out (" + _timeout + " msec)...");
                 $scope.loading = false;
                 }, _timeout);*/
            };

            $scope.stopLoading = function () {
                $timeout(function () {
                    $scope._loading = false;
                }, 500);
            };

            $scope.isOnline = function (player) {
                var playerID = player.facebookID;
                var state = onlinePlayers[playerID];
                if (!state) {
                    state = {connected: false};
                    onlinePlayers[playerID] = state;
                    $http.get("/api/online/" + playerID)
                        .success(function (newState) {
                            onlinePlayers[playerID] = newState;
                        })
                        .error(function (err) {
                            $log.error("Error retrieving online state for user " + playerID);
                        });

                }
                return state && state.connected;
            };

            $scope.range = function (n) {
                return new Array(n);
            };

            $scope.facebookLoginStatus = function (fbUserID) {
                $scope.postLoginUpdates(fbUserID, false);
            };

            $scope.login = function (event) {
                if (event) {
                    event.preventDefault();
                }
                Facebook.login().then(
                    function (response) {
                        var fbUserID = response.authResponse.userID;
                        $scope.postLoginUpdates(fbUserID, true);
                    },
                    function (err) {
                        $log.error("main:login err = " + angular.toJson(err));
                    });
            };

            $scope.logout = function (event) {
                if (event) {
                    event.preventDefault();
                }
                Facebook.logout();
                MySession.logout();
            };

            $scope.postLoginUpdates = function (fbUserID, userInitiated) {
                // capture the user ID
                MySession.fbUserID = fbUserID;

                // load the user's Facebook profile
                Facebook.getUserProfile().then(
                    function (response) {
                        MySession.fbProfile = response;
                        MySession.fbAuthenticated = true;
                    },
                    function (err) {
                        toaster.pop('error', 'Error!', "Facebook login error - " + err.data);
                    });

                // load the user's ShockTrade profile
                ProfileService.getProfileByFacebookID(fbUserID).then(
                    function (profile) {
                        if (!profile.error) {
                            $log.info("ShockTrade user profile loaded...");
                            MySession.userProfile = profile;
                            MySession.authenticated = true;

                            loadFacebookFriends();
                            $scope.filters = MySession.userProfile.filters;
                        }
                        else {
                            $log.info("Non-member identified... Launching Sign-up dialog...");
                            MySession.nonMember = true;
                            $scope.signUpPopup(fbUserID, MySession.fbProfile);
                        }
                    },
                    function (err) {
                        toaster.pop('error', 'Error!', "ShockTrade Profile retrieval error - " + err.data);
                        $scope.signUpPopup(fbUserID, MySession.fbProfile);
                    });
            };

            function loadFacebookFriends() {
                Facebook.getTaggableFriends().then(
                    function (response) {
                        var friends = response.data;
                        console.log(friends.length + " friends loaded");
                        MySession.fbFriends = friends.sort(function (a, b) {
                            if (a.name < b.name) return -1;
                            else if (a.name > b.name) return 1;
                            else return 0;
                        });
                    },
                    function (err) {
                        toaster.pop('error', 'Error!', "Failed to retrieve Facebook friends");
                    });
            }

            $scope.signUpPopup = function (fbUserID, fbProfile) {
                SignUpDialog.popup(fbUserID, fbProfile);
            };

            //////////////////////////////////////////////////////////////////////
            //              Quote-related Functions
            //////////////////////////////////////////////////////////////////////

            $scope.normalizeExchange = function (market) {
                if (market == null) return null;
                else {
                    var s = market.toUpperCase();
                    if (s.indexOf("ASE") == 0) return s;
                    else if (s.indexOf("CCY") == 0) return s;
                    else if (s.indexOf("NAS") == 0) return "NASDAQ";
                    else if (s.indexOf("NCM") == 0) return "NASDAQ";
                    else if (s.indexOf("NGM") == 0) return "NASDAQ";
                    else if (s.indexOf("NMS") == 0) return "NASDAQ";
                    else if (s.indexOf("NYQ") == 0) return "NYSE";
                    else if (s.indexOf("NYS") == 0) return "NYSE";
                    else if (s.indexOf("OBB") == 0) return "OTCBB";
                    else if (s.indexOf("OTC") == 0) return "OTCBB";
                    else if (s.indexOf("OTHER") == 0) return "OTHER_OTC";
                    else if (s.indexOf("PCX") == 0) return s;
                    else if (s.indexOf("PNK") == 0) return "OTCBB";
                    else {
                        $log.warn("exchange = " + s);
                        return s;
                    }
                }
            };

            $scope.getExchangeClass = function (exchange) {
                return $scope.normalizeExchange(exchange) + ' bold';
            };

            $scope.getPreferenceIcon = function (q) {
                // fail-safe
                if (!q || !q.symbol) return "";

                // check for favorite and held securities
                var symbol = q.symbol;
                if (HeldSecurities.isHeld(symbol)) return "fa fa-star";
                else if (FavoriteSymbols.isFavorite(symbol)) return "fa fa-heart";
                else return "";
            };

            //////////////////////////////////////////////////////////////////////
            //              Data Graphs
            //////////////////////////////////////////////////////////////////////

            // setup the tabs
            $scope.playTabs = [{
                "name": "Search",
                "icon_class": "fa-search",
                "tool_tip": "Search for games",
                "url": "/search",
                "isVisible": function () {
                    return true;
                }
            }, {
                "name": "Dashboard",
                "icon_class": "fa-gamepad",
                "tool_tip": "Main game dashboard",
                "url": "/dashboard",
                "isVisible": function () {
                    return !MySession.contestIsEmpty();
                }
            }, {
                "name": "Discover",
                "icon_class": "fa-newspaper-o",
                "tool_tip": "Stock News and Quotes",
                "url": "/discover",
                "isVisible": function () {
                    return true;
                }
            }, {
                "name": "Research",
                "icon_class": "fa-table",
                "tool_tip": "Stock Research",
                "url": "/research",
                "isVisible": function () {
                    return true;
                }
            }, {
                "name": "My Profile",
                "icon_class": "fa-user",
                "tool_tip": "My Profile",
                "url": "/profile",
                "isVisible": function () {
                    return MySession.authenticated;
                }
            }];

            // define the levels
            $scope.levels = [
                {"number": 1, "nextLevelXP": 1000, "description": "Private"},
                {"number": 2, "nextLevelXP": 2000, "description": "Private 1st Class"},
                {"number": 3, "nextLevelXP": 4000, "description": "Corporal"},
                {"number": 4, "nextLevelXP": 8000, "description": "First Corporal"},
                {"number": 5, "nextLevelXP": 16000, "description": "Sergeant"},
                {"number": 6, "nextLevelXP": 32000, "description": "Staff Sergeant"},
                {"number": 7, "nextLevelXP": 64000, "description": "Gunnery Sergeant"},
                {"number": 8, "nextLevelXP": 1280000, "description": "Master Sergeant"},
                {"number": 9, "nextLevelXP": 256000, "description": "First Sergeant"},
                {"number": 10, "nextLevelXP": 1024000, "description": "Sergeant Major"},
                {"number": 11, "nextLevelXP": 2048000, "description": "Warrant Officer 3rd Class"},
                {"number": 12, "nextLevelXP": 4096000, "description": "Warrant Officer 2nd Class"},
                {"number": 13, "nextLevelXP": 4096000, "description": "Warrant Officer 1st Class"},
                {"number": 14, "nextLevelXP": 8192000, "description": "Chief Warrant Officer"},
                {"number": 15, "nextLevelXP": 8192000, "description": "Master Chief Warrant Officer"},
                {"number": 16, "nextLevelXP": 16384000, "description": "Lieutenant"},
                {"number": 17, "nextLevelXP": 32768000, "description": "First Lieutenant"},
                {"number": 18, "nextLevelXP": 65536000, "description": "Captain"},
                {"number": 19, "nextLevelXP": 131072000, "description": "Major"},
                {"number": 20, "nextLevelXP": 262144000, "description": "Lieutenant Colonel"},
                {"number": 21, "nextLevelXP": 524288000, "description": "Colonel"},
                {"number": 22, "nextLevelXP": 524288000, "description": "Brigadier General"},
                {"number": 23, "nextLevelXP": 524288000, "description": "Major General"},
                {"number": 24, "nextLevelXP": 524288000, "description": "Lieutenant General"},
                {"number": 25, "nextLevelXP": 524288000, "description": "General"}];

            //////////////////////////////////////////////////////////////////////
            //              Event Listeners
            //////////////////////////////////////////////////////////////////////

            $scope.$on("user_status_changed", function (event, newState) {
                $log.info("user_status_changed: newState = " + angular.toJson(newState));
            });

            // watch for changes to the player's profile
            $scope.$watch("MySession.userProfile", function () {
                if (!MySession.userProfile.favorites) MySession.userProfile.favorites = ['AAPL'];
                if (!MySession.userProfile.recentSymbols) MySession.userProfile.recentSymbols = ['AAPL', 'AMZN', 'GOOG', 'MSFT'];

                // load the favorite and recent quotes
                FavoriteSymbols.setSymbols(MySession.userProfile.favorites);
                RecentSymbols.setSymbols(MySession.userProfile.recentSymbols);

                // load the held securities
                var id = MySession.getUserID();
                if (id) {
                    HeldSecurities.init(id);
                }
            });

        }]);
})();

File: app/client/angularjs/src/main/scala/com/shocktrade/client/MainController.scala
package com.shocktrade.client

import com.shocktrade.common.models.quote.ClassifiedQuote
import com.shocktrade.common.models.OnlineStatus
import com.shocktrade.client.MainController._
import com.shocktrade.client.ScopeEvents._
import com.shocktrade.client.contest.ContestService
import com.shocktrade.client.dialogs.SignUpDialog
import com.shocktrade.client.models.Profile
import com.shocktrade.client.profile.ProfileService
import org.scalajs.angularjs.facebook.FacebookService
import org.scalajs.angularjs.http.Http
import org.scalajs.angularjs.toaster._
import org.scalajs.angularjs.{Controller, Location, Scope, Timeout, injected, _}
import org.scalajs.dom.browser.console
import org.scalajs.nodejs.social.facebook.{FacebookProfileResponse, TaggableFriend}
import org.scalajs.nodejs.util.ScalaJsHelper._
import org.scalajs.sjs.JsUnderOrHelper._

import scala.concurrent.duration._
import scala.scalajs.concurrent.JSExecutionContext.Implicits.queue
import scala.scalajs.js
import scala.scalajs.js.JSON
import scala.scalajs.js.annotation.ScalaJSDefined
import scala.util.{Failure, Success}

/**
  * Main Controller
  * @author Lawrence Daniels <lawrence.daniels@gmail.com>
  */
class MainController($scope: MainControllerScope, $http: Http, $location: Location, $timeout: Timeout, toaster: Toaster,
                     @injected("ContestService") contestService: ContestService,
                     @injected("Facebook") facebook: FacebookService,
                     @injected("MySessionService") mySession: MySessionService,
                     @injected("ProfileService") profileService: ProfileService,
                     @injected("SignUpDialog") signUpDialog: SignUpDialog)
  extends Controller with GlobalLoading {

  private var loadingIndex = 0
  private var nonMember = true
  private val onlinePlayers = js.Dictionary[OnlineStatus]()

  $scope.appTabs = js.Array(
    new MainTab(name = "About", icon_class = "fa-info-circle", tool_tip = "About ShockTrade", url = "/about/us"),
    new MainTab(name = "Home", icon_class = "fa-home", tool_tip = "My Home page", url = "/home", authenticationRequired = true),
    new MainTab(name = "Search", icon_class = "fa-search", tool_tip = "Search for games", url = "/search"),
    new MainTab(name = "Dashboard", icon_class = "fa-gamepad", tool_tip = "Main game dashboard", url = "/dashboard", contestRequired = true),
    new MainTab(name = "Discover", icon_class = "fa-newspaper-o", tool_tip = "Stock News and Quotes", url = "/discover"),
    new MainTab(name = "Explore", icon_class = "fa-trello", tool_tip = "Explore Sectors and Industries", url = "/explore"),
    new MainTab(name = "Research", icon_class = "fa-database", tool_tip = "Stock Research", url = "/research"))

  $scope.levels = GameLevel.Levels

  ///////////////////////////////////////////////////////////////////////////
  //          Loading Functions
  ///////////////////////////////////////////////////////////////////////////

  $scope.isLoading = () => loadingIndex > 0

  $scope.startLoading = (timeout: js.UndefOr[Int]) => {
    loadingIndex += 1
    val _timeout = timeout getOrElse DEFAULT_TIMEOUT

    // set loading timeout
    $timeout(() => {
      console.log(s"Disabling the loading animation due to time-out (${_timeout} msec)...")
      loadingIndex = 0
    }, _timeout)
  }

  $scope.stopLoading = (promise: js.UndefOr[js.Promise[js.Any]]) => {
    $timeout.cancel(promise)
    $timeout(() => if (loadingIndex > 0) loadingIndex -= 1, 500.millis)
    ()
  }

  ///////////////////////////////////////////////////////////////////////////
  //          Public Functions
  ///////////////////////////////////////////////////////////////////////////

  $scope.mainInit = () => {
    console.log(s"Initializing ${getClass.getSimpleName}...")
  }

  $scope.getAssetCode = (q: js.UndefOr[ClassifiedQuote]) => MainController.getAssetCode(q)

  $scope.getAssetIcon = (q: js.UndefOr[ClassifiedQuote]) => MainController.getAssetIcon(q)

  $scope.getExchangeClass = (exchange: js.UndefOr[String]) => s"${normalizeExchange(exchange)} bold"

  $scope.getTabIndex = () => determineTableIndex

  $scope.isVisible = (aTab: js.UndefOr[ContestTab]) => aTab.exists { tab =>
    (loadingIndex == 0) && ((!tab.contestRequired || mySession.contest_?.isDefined) && (!tab.authenticationRequired || mySession.isAuthenticated))
  }

  $scope.normalizeExchange = (market: js.UndefOr[String]) => MainController.normalizeExchange(market)

  $scope.postLoginUpdates = (aFacebookID: js.UndefOr[String], aUserInitiated: js.UndefOr[Boolean]) => {
    for {
      facebookID <- aFacebookID
      userInitiated <- aUserInitiated
    } doPostLoginUpdates(facebookID, userInitiated)
  }

  //////////////////////////////////////////////////////////////////////
  //              My Session Service Functions
  //////////////////////////////////////////////////////////////////////

  $scope.contestIsEmpty = () => mySession.contest_?.isEmpty

  $scope.getContestID = () => mySession.getContestID

  $scope.getContestName = () => mySession.getContestName

  $scope.getContestStatus = () => mySession.getContestStatus

  $scope.getFacebookID = () => mySession.getFacebookID

  $scope.getFacebookProfile = () => mySession.getFacebookProfile

  $scope.getFacebookFriends = () => mySession.fbFriends_?

  $scope.getFundsAvailable = () => mySession.getFundsAvailable

  $scope.getNetWorth = () => mySession.getNetWorth.getOrElse(0d)

  $scope.getUserID = () => mySession.userProfile._id.orNull

  $scope.getUserName = () => mySession.getUserName.orNull

  $scope.getUserProfile = () => mySession.userProfile

  $scope.hasNotifications = () => mySession.hasNotifications

  $scope.hasPerk = (aPerkCode: js.UndefOr[String]) => {
    aPerkCode.exists(mySession.hasPerk)
  }

  $scope.isAdmin = () => mySession.isAdmin

  $scope.isAuthenticated = () => mySession.isAuthenticated

  //////////////////////////////////////////////////////////////////////
  //              Private Functions
  //////////////////////////////////////////////////////////////////////

  $scope.isOnline = (aPlayer: js.UndefOr[Profile]) => aPlayer.exists { player =>
    player._id.exists { playerID =>
      if (!onlinePlayers.contains(playerID)) {
        onlinePlayers(playerID) = new OnlineStatus(connected = false)
        profileService.getOnlineStatus(playerID) onComplete {
          case Success(newState) =>
            onlinePlayers(playerID) = newState
          case Failure(e) =>
            console.error(s"Error retrieving online state for user $playerID")
        }
      }
      onlinePlayers.get(playerID).exists(_.connected)
    }
  }

  $scope.getPreferenceIcon = (q: js.Dynamic) => {
    // fail-safe
    if (!isDefined(q) || !isDefined(q.symbol)) ""
    else {
      // check for favorite and held securities
      q.symbol.asOpt[String] map { symbol =>
        if (mySession.isFavoriteSymbol(symbol)) "fa fa-heart"
        //else if (heldSecurities.isHeld(symbol)) "fa fa-star"
        else ""
      } getOrElse ""
    }
  }

  $scope.login = () => {
    facebook.login() onComplete {
      case Success(response) =>
        nonMember = true

        // load the profile
        facebook.facebookID map (doPostLoginUpdates(_, userInitiated = true))
      case Failure(e) =>
        console.error(s"main:login error")
        e.printStackTrace()
    }
  }

  $scope.logout = () => {
    nonMember = false
    facebook.logout() onComplete {
      case Success(_) => mySession.logout()
      case Failure(e) =>
        toaster.error("An error occurred during logout")
        mySession.logout()
        e.printStackTrace()
    }
  }

  private def doPostLoginUpdates(facebookID: String, userInitiated: Boolean) = {
    console.log(s"facebookID = $facebookID, userInitiated = $userInitiated")

    // capture the Facebook user ID
    mySession.setFacebookID(facebookID)

    val outcome = for {
    // load the user"s Facebook profile
      fbProfile <- {
        console.log(s"Retrieving Facebook profile for FBID $facebookID...")
        facebook.getUserProfile
      }

      fbFriends <- {
        console.log(s"Loading Facebook friends for FBID $facebookID...")
        facebook.getTaggableFriends
      }

      // load the user"s ShockTrade profile
      profile <- {
        console.log(s"Retrieving ShockTrade profile for FBID $facebookID...")
        profileService.getProfileByFacebookID(facebookID)
      }
    } yield (fbProfile, fbFriends, profile)

    outcome onComplete {
      case Success((fbProfile, fbFriends, profile)) =>
        console.log("ShockTrade user profile, Facebook profile, and friends loaded...")
        nonMember = false
        mySession.setUserProfile(profile, fbProfile)
        mySession.fbFriends_? = fbFriends
      case Failure(e) =>
        toaster.error(s"ShockTrade Profile retrieval error - ${e.getMessage}")
    }
  }

  $scope.signUp = () => {
    signUpDialog.popup() onComplete {
      case Success((profile, fbProfile)) =>
        mySession.setUserProfile(profile, fbProfile)
      case Failure(e) =>
        toaster.error(e.getMessage)
    }
  }

  //////////////////////////////////////////////////////////////////////
  //              Tab Functions
  //////////////////////////////////////////////////////////////////////

  $scope.changeAppTab = (index: js.UndefOr[Int]) => index foreach { tabIndex =>
    mySession.userProfile._id.toOption match {
      case Some(userID) =>
        asyncLoading($scope)(profileService.setIsOnline(userID)) onComplete {
          case Success(outcome) =>
            if (isDefined(outcome.error)) {
              console.log(s"outcome = ${angular.toJson(outcome)}")
              toaster.error(outcome.error)
            }
            performTabSwitch(tabIndex)
          case Failure(e) =>
            toaster.error(e.getMessage)
        }
      case None =>
        performTabSwitch(tabIndex)
    }
  }

  private def performTabSwitch(tabIndex: Int): Unit = {
    val tab = $scope.appTabs(tabIndex)
    console.log(s"Changing location for ${mySession.getUserName} to ${tab.url}")
    $location.url(tab.url)
  }

  private def determineTableIndex: Int = $location.path() match {
    case path if path.contains("/about") => 0
    case path if path.contains("/home") => 1
    case path if path.contains("/search") => 2
    case path if path.contains("/dashboard") => 3
    case path if path.contains("/discover") => 4
    case path if path.contains("/explore") => 5
    case path if path.contains("/research") => 6
    case path => 0
  }

  //////////////////////////////////////////////////////////////////////
  //              Event Listeners
  //////////////////////////////////////////////////////////////////////

  $scope.onUserStatusChanged((_, newState) => console.log(s"user_status_changed: newState = ${JSON.stringify(newState)}"))

}

/**
  * Main Controller Singleton
  * @author Lawrence Daniels <lawrence.daniels@gmail.com>
  */
object MainController {
  private val DEFAULT_TIMEOUT = 15000

  protected[client] def getAssetCode(q: js.UndefOr[ClassifiedQuote]) = {
    q.flatMap(_.assetType) map {
      case "Crypto-Currency" => "&#xf15a" // fa-bitcoin
      case "Currency" => "&#xf155" // fa-dollar
      case "ETF" => "&#xf18d" // fa-stack-exchange
      case _ => "&#xf0ac" // fa-globe
    } getOrElse ""
  }

  protected[client] def getAssetIcon(q: js.UndefOr[ClassifiedQuote]) = {
    q.flatMap(_.assetType) map {
      case "Crypto-Currency" => "fa fa-bitcoin st_blue"
      case "Currency" => "fa fa-dollar st_blue"
      case "ETF" => "fa fa-stack-exchange st_blue"
      case _ => "fa fa-globe st_blue"
    } getOrElse "fa fa-globe st_blue"
  }

  protected[client] def normalizeExchange(aMarket: js.UndefOr[String]) = {
    aMarket.flat map (_.toUpperCase) map {
      //case s if s.contains("ASE") => s
      //case s if s.contains("CCY") => s
      case s if s.contains("NAS") => "NASDAQ"
      case s if s.contains("NCM") => "NASDAQ"
      case s if s.contains("NGM") => "NASDAQ"
      case s if s.contains("NMS") => "NASDAQ"
      case s if s.contains("NYQ") => "NYSE"
      case s if s.contains("NYS") => "NYSE"
      case s if s.contains("OBB") => "OTCBB"
      case s if s.contains("OTC") => "OTCBB"
      case s if s.contains("OTHER") => "OTHER_OTC"
      //case s if s.contains("PCX") => s
      case s if s.contains("PNK") => "OTCBB"
      case s => s
    } getOrElse ""
  }

  /**
    * Represents a Contest Tab
    * @author Lawrence Daniels <lawrence.daniels@gmail.com>
    */
  @js.native
  trait ContestTab extends js.Object {
    var contestRequired: Boolean = js.native
    var authenticationRequired: Boolean = js.native
  }

  /**
    * Represents a Main Tab
    * @author Lawrence Daniels <lawrence.daniels@gmail.com>
    */
  @ScalaJSDefined
  class MainTab(val name: String,
                val icon_class: String,
                val tool_tip: String,
                val url: String,
                val contestRequired: Boolean = false,
                val authenticationRequired: Boolean = false) extends js.Object

}

/**
  * Main Controller Scope
  * @author Lawrence Daniels <lawrence.daniels@gmail.com>
  */
@js.native
trait MainControllerScope extends Scope {
  // variables
  var appTabs: js.Array[MainTab] = js.native
  var levels: js.Array[GameLevel] = js.native

  // functions
  var isLoading: js.Function0[Boolean]
  var startLoading: js.Function1[js.UndefOr[Int], js.Promise[js.Any]] = js.native
  var stopLoading: js.Function1[js.UndefOr[js.Promise[js.Any]], Unit] = js.native

  var mainInit: js.Function0[Unit] = js.native
  var getAssetCode: js.Function1[js.UndefOr[ClassifiedQuote], String] = js.native
  var getAssetIcon: js.Function1[js.UndefOr[ClassifiedQuote], String] = js.native
  var getExchangeClass: js.Function1[js.UndefOr[String], String] = js.native
  var getTabIndex: js.Function0[Int] = js.native
  var isVisible: js.Function1[js.UndefOr[ContestTab], Boolean] = js.native
  var normalizeExchange: js.Function1[js.UndefOr[String], String] = js.native
  var postLoginUpdates: js.Function2[js.UndefOr[String], js.UndefOr[Boolean], Unit] = js.native

  var contestIsEmpty: js.Function0[Boolean] = js.native
  var getContestID: js.Function0[js.UndefOr[String]] = js.native
  var getContestName: js.Function0[String] = js.native
  var getContestStatus: js.Function0[String] = js.native
  var getFacebookID: js.Function0[js.UndefOr[String]] = js.native
  var getFacebookProfile: js.Function0[js.UndefOr[FacebookProfileResponse]] = js.native
  var getFacebookFriends: js.Function0[js.Array[TaggableFriend]] = js.native
  var getFundsAvailable: js.Function0[Double] = js.native
  var getNetWorth: js.Function0[Double] = js.native
  var getUserID: js.Function0[String] = js.native
  var getUserName: js.Function0[String] = js.native
  var getUserProfile: js.Function0[Profile] = js.native
  var hasNotifications: js.Function0[Boolean] = js.native
  var hasPerk: js.Function1[js.UndefOr[String], Boolean] = js.native
  var isAdmin: js.Function0[Boolean] = js.native
  var isAuthenticated: js.Function0[Boolean] = js.native

  var changeAppTab: js.Function1[js.UndefOr[Int], Unit] = js.native
  var isOnline: js.Function1[js.UndefOr[Profile], Boolean] = js.native
  var getPreferenceIcon: js.Function1[js.Dynamic, String] = js.native
  var login: js.Function0[Unit] = js.native
  var logout: js.Function0[Unit] = js.native
  var signUp: js.Function0[Unit] = js.native

}